Cache/Hashing.lean

```diff
@@ -84,41 +107,56 @@ Computes the hash of a file, which mixes:
 * The hash of its content
 * The hashes of the imported files that are part of `Mathlib`
 -/
-partial def getFileHash (filePath : FilePath) : HashM <| Option UInt64 := do
-  match (← get).cache[filePath]? with
+partial def getHash (mod : Name) (sourceFile : FilePath) : HashM <| Option UInt64 := do
+  match (← get).cache[mod]? with
   | some hash? => return hash?
   | none =>
-    let fixedPath := (← IO.getPackageDir filePath) / filePath
-    if !(← fixedPath.pathExists) then
-      IO.println s!"Warning: {fixedPath} not found. Skipping all files that depend on it."
-      if fixedPath.extension != "lean" then
-        IO.println s!"Note that `lake exe cache get ...` expects file names \
-          (e.g. `Mathlib/Init.lean`), not module names (e.g. `Mathlib.Init`)."
-      modify fun stt => { stt with cache := stt.cache.insert filePath none }
+    -- let fixedPath := (← IO.getPackageDir filePath) / filePath
+    if !(← sourceFile.pathExists) then
+      IO.println s!"Warning: {sourceFile} not found. Skipping all files that depend on it."
+      modify fun stt => { stt with cache := stt.cache.insert mod none }
       return none
-    let content ← IO.FS.readFile fixedPath
-    let fileImports := getFileImports content (← getPackageDirs)
+    let content ← IO.FS.readFile sourceFile
+    let fileImports ← getFileImports content mod
     let mut importHashes := #[]
-    for importHash? in ← fileImports.mapM getFileHash do
+    for importHash? in ← fileImports.mapM (fun imp => getHash imp.1 imp.2) do
       match importHash? with
       | some importHash => importHashes := importHashes.push importHash
       | none =>
-        modify fun stt => { stt with cache := stt.cache.insert filePath none }
+        -- one import did not have a hash --> invalidate hash of this module
+        modify fun stt => { stt with cache := stt.cache.insert mod none }
         return none
     let rootHash := (← get).rootHash
-    let pathHash := hash filePath.components
+    -- TODO: Ideally, we could hash the module name here, but since that
+    -- invalidates all existing cache, we recontruct this `c` which used to be
+    -- cached previously. Change this at an appropriate time!
```

Comment:
What would be an appropriate time?

Do we need to allow for some form of backwards compatible changes??

Reply:
if I understand correctly, it can be changed in this PR before merging and it just invalidates all of cache once.

But first I need to figure out why the CI fails so that it builds the new files. Currently it does something with `running batteries:optBarrel` which I dont understand why it does that.

Reply:
Actually, this change is basically equivalent to a lean-toolchain bump.

Reply:
Okay, thanks.